src/main/java/seedu/address/model/Concierge.java

```diff
@@ -182,76 +221,59 @@ public void checkInRoom(RoomNumber roomNumber) {
+        addCheckedInGuestIfNotPresent(guestToCheckIn);
     }
 
     /**
-     * Checks out a room's first booking using its room number and remove the guest from the checked-in guest list
+     * Checks out a room's first booking
      */
     public void checkoutRoom(RoomNumber roomNumber) {
         Room room = rooms.getRoom(roomNumber);
-        rooms.setRoom(room, room.checkout());
-
-        Booking bookingToCheckout = room.getBookings().getFirstBooking();
-        Guest guestToCheckout = bookingToCheckout.getGuest();
-
-        // If the guest exists in the archived guest list, this block does nothing. This is expected because a guest
-        // may have stayed in the hotel before, so he would already be in the archived guest list.
-        if (!hasGuest(guestToCheckout)) {
-            addGuest(guestToCheckout);
-        }
-
-        // If the guest does not exist in the checked-in guest list, or still has checked-in bookings in other rooms,
-        // this code below does nothing. This is expected
-        // behavior, because a guest can have multiple bookings at once and checkout one before another.
-        // Though not ideal, this is the current implementation.
-        // TODO Check if guest to remove from checked-in list still has other bookings in other rooms OR
-        // TODO Create new guest subclass that stores the room information, and add instances of that into
-        // checked-in guest list
-        if (!hasCheckedInGuest(guestToCheckout)
-                || rooms.asUnmodifiableObservableList().stream()
-                    .anyMatch(r -> r.getBookings().getSortedBookingsSet().stream()
-                        .anyMatch(b -> b.getIsCheckedIn() && b.getGuest().equals(guestToCheckout)))) {
-            return;
-        }
-        removeCheckedInGuest(guestToCheckout);
+        checkoutRoom(room, room.getBookings().getFirstBooking());
     }
 
     /**
-     * Checks out a room's booking using its room number and the specified booking period and remove the guest
-     * from the checked-in guest list
+     * Checks out a room's booking using the specified start date
      */
-    public void checkoutRoom(RoomNumber roomNumber, BookingPeriod bookingPeriod) {
+    public void checkoutRoom(RoomNumber roomNumber, LocalDate startDate) {
         Room room = rooms.getRoom(roomNumber);
-        rooms.setRoom(room, room.checkout(bookingPeriod));
+        checkoutRoom(room, room
+                .getBookings()
+                .getFirstBookingByPredicate(booking -> booking.getBookingPeriod().getStartDate().equals(startDate)));
+    }
+
+    /**
+     * Checks out the given booking from the given room.
+     *
+     * If the booking's guest was checked-in (i.e. exists in the checked-in guest list), check:
+     * 1) If the guest does not have checked-in bookings in other rooms, remove him from the checked-in guest list.
+     * 2) If the guest exists in the archived guest list. add him to the archived guest list
```

Comment:
This may be a typo --> "does not exist" instead of "exists"?

Reply:
Yup, typo. Will fix.